src/libraries/System.Formats.Asn1/src/System/Formats/Asn1/AsnDecoder.BitString.cs

```diff
@@ -0,0 +1,841 @@
+        /// <param name="expectedTag">
+        ///   The tag to check for before reading, or <see langword="null"/> for the default tag (Universal 3).
+        /// </param>
+        /// <returns>
+        ///   An array containing the contents of the Bit String value.
+        /// </returns>
+        /// <remarks>
+        ///   The least significant bits in the last byte which are reported as "unused" by the
+        ///   <paramref name="unusedBitCount"/> value will be copied into the return value
+        ///   as unset bits, irrespective of their value in the encoded representation.
+        /// </remarks>
+        /// <exception cref="ArgumentOutOfRangeException">
+        ///   <paramref name="ruleSet"/> is not defined.
+        /// </exception>
+        /// <exception cref="AsnContentException">
+        ///   the next value does not have the correct tag.
+        ///
+        ///   -or-
+        ///
+        ///   the length encoding is not valid under the current encoding rules.
+        ///
+        ///   -or-
+        ///
+        ///   the contents are not valid under the current encoding rules.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        ///   <paramref name="expectedTag"/>.<see cref="Asn1Tag.TagClass"/> is
+        ///   <see cref="TagClass.Universal"/>, but
+        ///   <paramref name="expectedTag"/>.<see cref="Asn1Tag.TagValue"/> is not correct for
+        ///   the method.
+        /// </exception>
+        /// <seealso cref="TryReadPrimitiveBitString"/>
+        /// <seealso cref="TryReadBitString"/>
+        public static byte[] ReadBitString(
+            ReadOnlySpan<byte> source,
+            AsnEncodingRules ruleSet,
+            out int unusedBitCount,
+            out int bytesConsumed,
+            Asn1Tag? expectedTag = null)
+        {
+            if (TryReadPrimitiveBitStringCore(
+                source,
+                ruleSet,
+                expectedTag ?? Asn1Tag.PrimitiveBitString,
+                out int? contentsLength,
+                out int headerLength,
+                out int localUbc,
+                out ReadOnlySpan<byte> localValue,
+                out int consumed,
+                out byte normalizedLastByte))
+            {
+                byte[] ret = localValue.ToArray();
+
+                // Update the last byte in case it's a non-canonical byte in a BER encoding.
+                if (localValue.Length > 0)
+                {
+                    ret[ret.Length - 1] = normalizedLastByte;
+                }
+
+                unusedBitCount = localUbc;
+                bytesConsumed = consumed;
+                return ret;
+            }
+
+            // If we get here, the tag was appropriate, but the encoding was constructed.
+
+            // Guaranteed long enough
+            int tooBig = contentsLength ?? SeekEndOfContents(source.Slice(headerLength), ruleSet);
+
+            byte[] rented = CryptoPool.Rent(tooBig);
```

Comment:
Possible untrusted length being passed to allocator. Please double-check.

Flow: `ReadPrimitiveBitStringCore` encounters a constructed tag in the string and the current rule set is BER. This will cause `ReadTagAndLength`'s normal exception handling logic not to run, outing a payload-controlled non-null _contentsLength_ value. `TryReadPrimitiveBitStringCore` will also early-exit, causing its `Slice` bounds checks not to run and returning _false_. Now, since _contentsLength_ is not null, an allocation will take place with a payload-controlled capacity.

Same question in the method `GetOctetStringContents` (AsnDecoder.OctetString.cs, lines 495 - 505), which calls this helper method similarly.